doc/src/tutorial/codegen.rst

```diff
@@ -0,0 +1,516 @@
+    #endif
+
+Various flags to ``codegen`` let you modify things. The project name for preprocessor 
+instructions can be varied using ``project``. Variables listed as global variables in 
+arg ``global_vars`` will not show up as function arguments.
+
+``language`` is a case-insensitive string that indicates the source code language. 
+Currently, 'C', 'F95' and 'Octave' are supported. 
+'Octave' generates code compatible with both Octave and Matlab.
+
+``header`` when True, a header is written on top of each source file. ``empty`` 
+when True, empty lines are used to structure the code. With ``argument_sequence``
+a sequence of arguments for the routine can be defined in a preferred order.  
+
+``prefix`` defines a prefix for the names of the files that contain the source code. 
+If omitted, the name of the first name_expr tuple is used.
+``to_files`` when True, the code will be written to one or more files with the
+given prefix.
+          
+
+
+Here is an example::
+
+    >>> [(f_name, f_code), header] = codegen(("volume",length*breadth*height), "F95",header=True, empty=False, argument_sequence=(breadth, length), global_vars=(height,))
+    >>> print(f_code)
+    !******************************************************************************
+    !*                    Code generated with sympy 0.7.7.dev                     *
+    !*                                                                            *
+    !*              See http://www.sympy.org/ for more information.               *
+    !*                                                                            *
+    !*                       This file is part of 'project'                       *
+    !******************************************************************************
+    REAL*8 function volume(breadth, length)
+    implicit none
+    REAL*8, intent(in) :: breadth
+    REAL*8, intent(in) :: length
+    volume = breadth*height*length
+    end function
+
+
+
+The method ``make_routine`` creates a ``Routine`` object, which represents an evaluation
+routine for a set of expressions. This is only good for internal use by the CodeGen 
+objects, as an intermediate representation from SymPy expression to generated code. 
+It is not recommended to make a ``Routine`` object yourself. You should instead use 
+``make_routine`` method. ``make_routine`` in turn calls the ``routine`` method of 
+the CodeGen object depending upon the language of choice. This creates the internal 
+objects representing assignments and so on, and creates the ``Routine`` class with them.
+
+The various codegen objects such as ``Routine`` and ``Variable`` aren't SymPy 
+objects (they don't subclass from Basic).
+
+For example::
+
+    >>> from sympy.utilities.codegen import make_routine
+    >>> from sympy.physics.hydrogen import R_nl
+    >>> from sympy import symbols, init_printing
+    >>> init_printing()
+    >>> x, y = symbols('x y')
+    >>> expr = R_nl(3,y,x,6)
```

Comment:
PEP8 > Spaces after the commas.

Please have a look at PEP8 and/or run a linter over your code to check PEP8 compliance. Please go through this whole PR and fix the PEP8 errors.